File: TorchOpt/_src/accelerated_op/adam_op/AdamOp.py
# Copyright 2022 MetaOPT Team. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from typing import Any
import torch
from TorchOpt._lib import adam_op


class AdamOp(object):
    class MuOp(torch.autograd.Function):
        @staticmethod
        def jvp(ctx: Any, *grad_inputs: Any) -> Any:
            pass

        @staticmethod
        def forward(ctx, *args):
            updates, mu, b1 = args
            new_mu = adam_op.forwardMu(updates, mu, b1)
            ctx.save_for_backward(updates, mu)
            ctx.b1 = b1
            return new_mu

        @staticmethod
        def backward(ctx, *args):
            dmu = args[0]
            updates, mu = ctx.saved_tensors
            b1 = ctx.b1
            result = adam_op.backwardMu(dmu, updates, mu, b1)
            return result[0], result[1], None

    class NuOp(torch.autograd.Function):
        @staticmethod
        def jvp(ctx: Any, *grad_inputs: Any) -> Any:
            pass

        @staticmethod
        def forward(ctx, *args):
            updates, nu, b2 = args
            new_nu = adam_op.forwardNu(updates, nu, b2)
            ctx.save_for_backward(updates, nu)
            ctx.b2 = b2
            return new_nu

        @staticmethod
        def backward(ctx, *args):
            dnu = args[0]
            updates, nu = ctx.saved_tensors
            b2 = ctx.b2
            result = adam_op.backwardNu(dnu, updates, nu, b2)
            return result[0], result[1], None

    class UpdatesOp(torch.autograd.Function):
        @staticmethod
        def jvp(ctx: Any, *grad_inputs: Any) -> Any:
            pass

        @staticmethod
        def forward(ctx, *args):
            new_mu, new_nu, (b1, b2, eps, eps_root, count) = args
            new_updates = adam_op.forwardUpdates(
                new_mu, new_nu, b1, b2, eps, eps_root, count)
            ctx.save_for_backward(new_updates, new_mu, new_nu)
            ctx.others = (b1, b2, eps, eps_root, count)
            return new_updates

        @staticmethod
        def backward(ctx, *args):
            dupdates = args[0]
            updates, new_mu, new_nu = ctx.saved_tensors
            b1, b2, eps, eps_root, count = ctx.others
            result = adam_op.backwardUpdates(
                dupdates, updates, new_mu, new_nu, b1, b2, count)
            return result[0], result[1], None

    def __init__(self, b1=0.9, b2=0.999, eps=1e-8, eps_root=0., inplace=True):
        self.b1 = b1
        self.b2 = b2
        self.eps = eps
        self.eps_root = eps_root
        self.inplace = inplace

    def __call__(self, mu, nu, updates, count):
        if updates is None:
            return mu, nu, None
        if self.inplace:
            new_updates, new_mu, new_nu = adam_op.forward_(updates, mu, nu, self.b1,
                                                           self.b2, self.eps, self.eps_root, count)
        else:
            new_mu = self.MuOp.apply(updates, mu, self.b1)
            new_nu = self.NuOp.apply(updates, nu, self.b2)
            new_updates = self.UpdatesOp.apply(
                new_mu,
                new_nu,
                (self.b1, self.b2, self.eps, self.eps_root, count))
        return new_mu, new_nu, new_updates

File: src/adam_op/adam_op.cpp
// Copyright 2022-2023 MetaOPT Team. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================

#include "include/adam_op/adam_op.h"

#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include "include/adam_op/adam_op_impl_cpu.h"
#if defined(__USE_CUDA__)
#include "include/adam_op/adam_op_impl_cuda.cuh"
#endif

namespace torchopt {

namespace py = pybind11;

namespace adam_op {

TensorArray<3> adamForwardInplace(const torch::Tensor &updates,
                                  const torch::Tensor &mu,
                                  const torch::Tensor &nu,
                                  const pyfloat_t b1,
                                  const pyfloat_t b2,
                                  const pyfloat_t eps,
                                  const pyfloat_t eps_root,
                                  const pyuint_t count) {
#if defined(__USE_CUDA__)
  if (updates.device().is_cuda()) {
    return adamForwardInplaceCUDA(updates, mu, nu, b1, b2, eps, eps_root, count);
  }
#endif
  if (updates.device().is_cpu()) {
    return adamForwardInplaceCPU(updates, mu, nu, b1, b2, eps, eps_root, count);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

torch::Tensor adamForwardMu(const torch::Tensor &updates,
                            const torch::Tensor &mu,
                            const pyfloat_t b1) {
#if defined(__USE_CUDA__)
  if (updates.device().is_cuda()) {
    return adamForwardMuCUDA(updates, mu, b1);
  }
#endif
  if (updates.device().is_cpu()) {
    return adamForwardMuCPU(updates, mu, b1);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

torch::Tensor adamForwardNu(const torch::Tensor &updates,
                            const torch::Tensor &nu,
                            const pyfloat_t b2) {
#if defined(__USE_CUDA__)
  if (updates.device().is_cuda()) {
    return adamForwardNuCUDA(updates, nu, b2);
  }
#endif
  if (updates.device().is_cpu()) {
    return adamForwardNuCPU(updates, nu, b2);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

torch::Tensor adamForwardUpdates(const torch::Tensor &new_mu,
                                 const torch::Tensor &new_nu,
                                 const pyfloat_t b1,
                                 const pyfloat_t b2,
                                 const pyfloat_t eps,
                                 const pyfloat_t eps_root,
                                 const pyuint_t count) {
#if defined(__USE_CUDA__)
  if (new_mu.device().is_cuda()) {
    return adamForwardUpdatesCUDA(new_mu, new_nu, b1, b2, eps, eps_root, count);
  }
#endif
  if (new_mu.device().is_cpu()) {
    return adamForwardUpdatesCPU(new_mu, new_nu, b1, b2, eps, eps_root, count);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

TensorArray<2> adamBackwardMu(const torch::Tensor &dmu,
                              const torch::Tensor &updates,
                              const torch::Tensor &mu,
                              const pyfloat_t b1) {
#if defined(__USE_CUDA__)
  if (dmu.device().is_cuda()) {
    return adamBackwardMuCUDA(dmu.contiguous(), updates, mu, b1);
  }
#endif
  if (dmu.device().is_cpu()) {
    return adamBackwardMuCPU(dmu.contiguous(), updates, mu, b1);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

TensorArray<2> adamBackwardNu(const torch::Tensor &dnu,
                              const torch::Tensor &updates,
                              const torch::Tensor &nu,
                              const pyfloat_t b2) {
#if defined(__USE_CUDA__)
  if (dnu.device().is_cuda()) {
    return adamBackwardNuCUDA(dnu.contiguous(), updates, nu, b2);
  }
#endif
  if (dnu.device().is_cpu()) {
    return adamBackwardNuCPU(dnu.contiguous(), updates, nu, b2);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

TensorArray<2> adamBackwardUpdates(const torch::Tensor &dupdates,
                                   const torch::Tensor &updates,
                                   const torch::Tensor &new_mu,
                                   const torch::Tensor &new_nu,
                                   const pyfloat_t b1,
                                   const pyfloat_t b2,
                                   const pyfloat_t eps_root,
                                   const pyuint_t count) {
#if defined(__USE_CUDA__)
  if (dupdates.device().is_cuda()) {
    return adamBackwardUpdatesCUDA(
        dupdates.contiguous(), updates, new_mu, new_nu, b1, b2, eps_root, count);
  }
#endif
  if (dupdates.device().is_cpu()) {
    return adamBackwardUpdatesCPU(
        dupdates.contiguous(), updates, new_mu, new_nu, b1, b2, eps_root, count);
  } else {
    throw std::runtime_error("Not implemented");
  }
}

void buildSubmodule(py::module &mod) {  // NOLINT[runtime/references]
  py::module m = mod.def_submodule("adam_op", "Adam Ops");
  m.def("forward_",
        &adamForwardInplace,
        "Adam forward inplace",
        py::arg("updates"),
        py::arg("mu"),
        py::arg("nu"),
        py::arg("b1"),
        py::arg("b2"),
        py::arg("eps"),
        py::arg("eps_root"),
        py::arg("count"));
  m.def("forward_mu",
        &adamForwardMu,
        "Adam forward mu",
        py::arg("updates"),
        py::arg("mu"),
        py::arg("b1"));
  m.def("forward_nu",
        &adamForwardNu,
        "Adam forward nu",
        py::arg("updates"),
        py::arg("nu"),
        py::arg("b2"));
  m.def("forward_updates",
        &adamForwardUpdates,
        "Adam forward updates",
        py::arg("new_mu"),
        py::arg("new_nu"),
        py::arg("b1"),
        py::arg("b2"),
        py::arg("eps"),
        py::arg("eps_root"),
        py::arg("count"));
  m.def("backward_mu",
        &adamBackwardMu,
        "Adam backward mu",
        py::arg("dmu"),
        py::arg("updates"),
        py::arg("mu"),
        py::arg("b1"));
  m.def("backward_nu",
        &adamBackwardNu,
        "Adam backward nu",
        py::arg("dnu"),
        py::arg("updates"),
        py::arg("nu"),
        py::arg("b1"));
  m.def("backward_updates",
        &adamBackwardUpdates,
        "Adam backward updates",
        py::arg("dupdates"),
        py::arg("updates"),
        py::arg("new_mu"),
        py::arg("new_nu"),
        py::arg("b1"),
        py::arg("b2"),
        py::arg("eps_root"),
        py::arg("count"));
}

}  // namespace adam_op
}  // namespace torchopt
